lib/Artwork.php

```diff
@@ -227,12 +241,102 @@ public static function GetByUrlPath($artistUrlName, $artworkUrlName): ?Artwork{
 		return $result[0];
 	}
 
+	public static function Build(string $artistName, ?int $artistDeathYear, string $artworkName, ?int $completedYear,
+				     bool $completedYearIsCirca, ?string $artworkTags, ?int $publicationYear,
+				     ?string $publicationYearPage, ?string $copyrightPage, ?string $artworkPage,
+				     ?string $museumPage): Artwork{
+		/** @return array<ArtworkTag> */
+		function parseArtworkTags(?string $artworkTags): array{
+			if (!$artworkTags) return array();
+
+			$artworkTags = array_map('trim', explode(',', $artworkTags)) ?? array();
+			$artworkTags = array_values(array_filter($artworkTags)) ?? array();
+			$artworkTags = array_unique($artworkTags);
+
+			return array_map(function ($str){
+				$artworkTag = new ArtworkTag();
+				$artworkTag->Name = $str;
+				return $artworkTag;
+			}, $artworkTags);
+		}
+
+		$artist = new Artist();
+		$artist->Name = $artistName;
+		$artist->DeathYear = $artistDeathYear;
+
+		$artwork = new Artwork();
+		$artwork->Artist = $artist;
+		$artwork->Name = $artworkName;
+		$artwork->CompletedYear = $completedYear;
+		$artwork->CompletedYearIsCirca = $completedYearIsCirca;
+		$artwork->ArtworkTags = parseArtworkTags($artworkTags);
+		$artwork->Status = 'unverified';
+		$artwork->Created = new DateTime();
+		$artwork->PublicationYear = $publicationYear;
+		$artwork->PublicationYearPage = $publicationYearPage;
+		$artwork->CopyrightPage = $copyrightPage;
+		$artwork->ArtworkPage = $artworkPage;
+		$artwork->MuseumPage = $museumPage;
+
+		return $artwork;
+	}
+
 	/**
 	 * @throws \Exceptions\ValidationException
+	 * @throws \Exceptions\InvalidImageUploadException
 	 */
-	public function Create(): void{
+	public function Create(string $uploadPath): void{
 		$this->Validate();
-		$this->Created = new DateTime();
+		$this->ValidateImageUpload($uploadPath);
+
+		$thumbnail = self::GenerateThumbnail($uploadPath);
+
+		/** @var ArtworkTag $artworkTag */
+		foreach ($this->ArtworkTags as $artworkTag) {
+			$artworkTag->GetOrCreate();
+		}
+
+		$this->Artist->GetOrCreate();
+		$this->Insert();
+
+		$storedImage = move_uploaded_file($uploadPath, WEB_ROOT . $this->ImageUrl);
+		$storedThumb = imagejpeg($thumbnail, WEB_ROOT . $this->ThumbUrl);
+
+		if (!($storedImage && $storedThumb)){
+			$log = new Log(ARTWORK_UPLOADS_LOG_FILE_PATH);
+			$log->Write("Failed to store image or thumbnail for uploaded artwork [$this->ArtworkId]");
+		}
```

Comment:
Alex's [comment](https://github.com/standardebooks/web/issues/234#issuecomment-1618939776) in #234 about deleting incomplete state applies here. If one of these two file operations failed, we're in an incomplete state, so we should delete the orphan DB record created on line 300 above. The log statement is fine, but the code should also throw an exception to `post.php`.

Reply:
It's not necessary to create a DB record before we validate/manipulate images. Images can be validated first, and generated thumbs can go in temp files. Then after they're generated and validated, and everything else in the object passes validation, we create the DB record and then move all of the images into their final filenames which will be based on the DB primary key.

Reply:
OK, I've updated the `Create` method. It now does the following;

1. Validate
2. Create both images & store in temporary files (in the actual `images/uploads` directory)
3. Create database objects
4. Rename temporary files

The error handling on the final step consists of logging and warning the user that something went wrong. I think this is acceptable because a) there aren't really any realistic reasons why it would be possible to create the files but not rename them[^1], so this should never happen, and b) if it does happen, everything required to recover is available in the db, filesystem and logs.

[^1]: maybe if a file has already been created and had its permissions changed, which would have to have been done manually.